[ContributorDoc] Checklist for adding new LSP/BSP messages (#2626)
apple/sourcekit-lsp: Contributor Documentation/Adding LSP and BSP Messages.md

# Adding LSP and BSP Messages

This document lists every file that must be updated when adding a new LSP or BSP request or notification.

## LSP Message

### In `swift-tools-protocols`

- [ ] Add the Swift file under `Sources/LanguageServerProtocol/Requests/` *(request)* or `Sources/LanguageServerProtocol/Notifications/` *(notification)*
- [ ] `Sources/LanguageServerProtocol/CMakeLists.txt` — add the file in alphabetical order
- [ ] `Sources/LanguageServerProtocol/Messages.swift` — add to `builtinRequests` *(request)* or `builtinNotifications` *(notification)*

### In `sourcekit-lsp`

- [ ] `Sources/SourceKitLSP/MessageHandlingDependencyTracker.swift` — add a case to `init(_ request:)` *(request)* or `init(_ notification:)` *(notification)* to declare the dependency type (`.freestanding`, `.documentRequest(uri)`, etc.) *(client → server only)*
- [ ] `Sources/SourceKitLSP/SourceKitLSPServer.swift`:
  - [ ] *(client → server)* Add a case to the main request dispatch switch *(request)* or `handle(notification:)` *(notification)* and implement the handler
  - [ ] Advertise via `serverCapabilities()`

>   - [ ] If handled by a language service: also advertise in each `LanguageService.initialize` conforming method

Reviewer: I mean, yes it’s going to be handled by a language service, otherwise we wouldn’t do anything in SourceKit-LSP, right?

Author: Not all requests are handled by LanguageService. E.g. `WorkspaceSymbolNames` (#2619) is not forwarded to any LanguageService. Am I missing something?

  - [ ] If experimental *(server → client)*: check the client capability before sending
- [ ] `Sources/SourceKitLSP/CapabilityRegistry.swift` — if gated on an experimental **client** capability, add a helper property

> - [ ] `Contributor Documentation/LSP Extensions.md` — document if it is a SourceKit-LSP extension

Reviewer: ```suggestion
- [ ] `Contributor Documentation/LSP Extensions.md` — document if it is an LSP extension
```

## BSP Message

### In `swift-tools-protocols`

- [ ] Add the Swift file under `Sources/BuildServerProtocol/Messages/`
- [ ] `Sources/BuildServerProtocol/CMakeLists.txt` — add the file in alphabetical order
- [ ] `Sources/BuildServerProtocol/Messages.swift` — add to `requestTypes` *(request)* or `notificationTypes` *(notification)*

### In `sourcekit-lsp`

- [ ] `Sources/BuildServerIntegration/BuildServerMessageDependencyTracker.swift` — add a case to the dependency tracking switch *(request only)*
- [ ] `Sources/BuildServerIntegration/BuiltInBuildServer.swift` — add the handler method to the `BuiltInBuildServer` protocol *(request only)*
- [ ] `Sources/BuildServerIntegration/BuiltInBuildServerAdapter.swift` — add a case to the request dispatch switch *(request only)*

> - [ ] *(SourceKit-LSP → build server, notification only)* Send via `buildServerManager.send(notification)` from the appropriate place; if built-in servers also need to react, add a case to `BuiltInBuildServerAdapter.handle(notification:)`

Reviewer: Similar to my comment in LSP: I don’t think we need to document that you need to send the request if you want to send it.

- [ ] *(build server → SourceKit-LSP, notification only)* `Sources/BuildServerIntegration/BuildServerManager.swift` — add a case to `handle(notification:)` and implement the handler

> - [ ] `Contributor Documentation/BSP Extensions.md` — document if it is a SourceKit-LSP extension

Reviewer: ```suggestion
- [ ] `Contributor Documentation/BSP Extensions.md` — document if it is a BSP extension
```

# Adding LSP and BSP Messages

This document lists every file that must be updated when adding a new LSP or BSP request or notification.

## LSP Request

### In `swift-tools-protocols`

- [ ] Add the request Swift file under `Sources/LanguageServerProtocol/Requests/`
- [ ] `Sources/LanguageServerProtocol/CMakeLists.txt` — add the file in alphabetical order

> - [ ] `Sources/LanguageServerProtocol/Messages.swift` — add to `builtinRequests`

Author: This tend to be missed because our test cases don't go through this. We should fix that...

  - [ ] Add a case to the main request dispatch switch
  - [ ] Implement the handler function
  - [ ] If experimental: advertise via `serverCapabilities()` (add an entry to `experimentalCapabilities`) and guard the handler with a capability check
- [ ] `Sources/SourceKitLSP/CapabilityRegistry.swift` — if the request is gated on an experimental **client** capability, add a helper property
- [ ] `Contributor Documentation/LSP Extensions.md` — document the request
- [ ] Add tests

## LSP Notification

### In `swift-tools-protocols`

- [ ] Add the notification Swift file under `Sources/LanguageServerProtocol/Notifications/`
- [ ] `Sources/LanguageServerProtocol/CMakeLists.txt` — add the file in alphabetical order
- [ ] `Sources/LanguageServerProtocol/Messages.swift` — add to `builtinNotifications`

### In `sourcekit-lsp`

- [ ] `Sources/SourceKitLSP/MessageHandlingDependencyTracker.swift` — add a case to the `init(_ notification:)` switch
- [ ] `Sources/SourceKitLSP/SourceKitLSPServer.swift` — add a case to `handle(notification:)` and implement the handler
- [ ] `Contributor Documentation/LSP Extensions.md` — document the notification
- [ ] Add tests

## BSP Request

> BSP requests are sent **from SourceKit-LSP (BSP client) to BSP servers**. Built-in BSP servers conform to the `BuiltInBuildServer` protocol; external BSP servers communicate over JSON-RPC via `ExternalBuildServerAdapter`.

Author: This isn't really true. There're server-to-client requests.

### In `swift-tools-protocols`

- [ ] Add the request Swift file under `Sources/BuildServerProtocol/Messages/`
- [ ] `Sources/BuildServerProtocol/CMakeLists.txt` — add the file in alphabetical order
- [ ] `Sources/BuildServerProtocol/Messages.swift` — add to `requestTypes` (registered as `MessageRegistry.bspProtocol`)

### In `sourcekit-lsp`

- [ ] `Sources/BuildServerIntegration/BuildServerMessageDependencyTracker.swift` — add a case to the dependency tracking switch
- [ ] `Sources/BuildServerIntegration/BuiltInBuildServer.swift` — add the handler method to the `BuiltInBuildServer` protocol
- [ ] `Sources/BuildServerIntegration/BuiltInBuildServerAdapter.swift` — add a case to the request dispatch switch that calls the protocol method
- [ ] Implement the handler in all `BuiltInBuildServer` conformers:
  - [ ] `SwiftPMBuildServer.swift`
  - [ ] `FixedCompilationDatabaseBuildServer.swift`
  - [ ] `JSONCompilationDatabaseBuildServer.swift`
- [ ] `Contributor Documentation/BSP Extensions.md` — document the request if it is a SourceKit-LSP extension
- [ ] Add tests

## BSP Notification

> BSP notifications flow **from BSP servers to SourceKit-LSP**. `BuildServerManager` is the central handler for all incoming BSP notifications.

Author: Ditto, there're client-to-server notifications.

# Adding LSP and BSP Messages

This document lists every file that must be updated when adding a new LSP or BSP request or notification.

## LSP Request

### In `swift-tools-protocols`

- [ ] Add the request Swift file under `Sources/LanguageServerProtocol/Requests/`
- [ ] `Sources/LanguageServerProtocol/CMakeLists.txt` — add the file in alphabetical order
- [ ] `Sources/LanguageServerProtocol/Messages.swift` — add to `builtinRequests`

### In `sourcekit-lsp`

- [ ] `Sources/SourceKitLSP/MessageHandlingDependencyTracker.swift` — add a case to `init(_ request:)` to declare the dependency type (`.freestanding`, `.documentRequest(uri)`, etc.) *(client → server only)*
- [ ] `Sources/SourceKitLSP/SourceKitLSPServer.swift`:
  - [ ] *(client → server)* Add a case to the main request dispatch switch and implement the handler

>   - [ ] *(server → client)* Call `client.send(request)` from the appropriate place

Reviewer: I’m not sure if we need this. I mean, if you want to use it, you have to send the request, otherwise there’s just no point.

>   - [ ] If experimental: advertise via `serverCapabilities()` and guard with a capability check *(client → server)*; check the client capability before sending *(server → client)*

Reviewer: We need to advertise the request from both `SourceKitLSPServer.serverCapabilities` and `SwiftLangaugeService.initialize`.

- [ ] `Sources/SourceKitLSP/CapabilityRegistry.swift` — if gated on an experimental **client** capability, add a helper property *(client → server only)*
- [ ] `Contributor Documentation/LSP Extensions.md` — document the request

> - [ ] Add tests

Reviewer: I don’t think we should explicitly need to mention this. Tests should be added to every change, no matter what.

> ## LSP Notification

Reviewer: Should we unify this with the `LSP Requests` section? Most of it is duplicated.

  - [ ] *(server → client)* Call `sendNotificationToClient(notification)` from the appropriate place; check the client capability before sending
- [ ] `Contributor Documentation/LSP Extensions.md` — document the notification
- [ ] Add tests

## BSP Request

BSP requests are sent **from SourceKit-LSP (BSP client) to BSP servers**. Built-in BSP servers conform to the `BuiltInBuildServer` protocol; external servers communicate via `ExternalBuildServerAdapter` over JSON-RPC and require no additional code changes.

### In `swift-tools-protocols`

- [ ] Add the request Swift file under `Sources/BuildServerProtocol/Messages/`
- [ ] `Sources/BuildServerProtocol/CMakeLists.txt` — add the file in alphabetical order
- [ ] `Sources/BuildServerProtocol/Messages.swift` — add to `requestTypes`

### In `sourcekit-lsp`

- [ ] `Sources/BuildServerIntegration/BuildServerMessageDependencyTracker.swift` — add a case to the dependency tracking switch
- [ ] `Sources/BuildServerIntegration/BuiltInBuildServer.swift` — add the handler method to the `BuiltInBuildServer` protocol
- [ ] `Sources/BuildServerIntegration/BuiltInBuildServerAdapter.swift` — add a case to the request dispatch switch

> - [ ] Implement the handler in all `BuiltInBuildServer` conformers:

Reviewer: I don’t think we need to explicitly spell this out. Adding the requirement to `BuiltInBuildServer` will prompt theses.
